kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/AwsV4SigningHttpRequestBuilder.java

```diff
@@ -0,0 +1,364 @@
+        builder.timeout(duration);
+        return this;
+    }
+
+    @Override
+    public Builder setHeader(String name, String value) {
+        builder.setHeader(name, value);
+        return this;
+    }
+
+    @Override
+    public Builder GET() {
+        builder.GET();
+        return this;
+    }
+
+    @Override
+    public Builder POST(BodyPublisher bodyPublisher) {
+        builder.POST(digestingPublisher(bodyPublisher));
+        return this;
+    }
+
+    @Override
+    public Builder PUT(BodyPublisher bodyPublisher) {
+        builder.PUT(digestingPublisher(bodyPublisher));
+        return this;
+    }
+
+    @Override
+    public Builder DELETE() {
+        builder.DELETE();
+        return this;
+    }
+
+    @Override
+    public Builder method(String method, BodyPublisher bodyPublisher) {
+        builder.method(method, digestingPublisher(bodyPublisher));
+        return this;
+    }
+
+    @Override
+    public Builder copy() {
+        return new AwsV4SigningHttpRequestBuilder(accessKey, secretKey, region, service, date, builder.copy());
+    }
+
+    @Override
+    public Builder uri(URI uri) {
+        builder.uri(uri);
+        return this;
+    }
+
+    @Override
+    public HttpRequest build() {
+        signRequest();
+        return builder.build();
+    }
+
+    @NonNull
+    private BodyPublisher digestingPublisher(BodyPublisher bodyPublisher) {
+        var items = new ArrayList<BodyPublisher>();
+
+        bodyPublisher.subscribe(new Flow.Subscriber<>() {
+            final MessageDigest digest = getSha256Digester();
+
+            @Override
+            public void onSubscribe(Flow.Subscription subscription) {
+                subscription.request(Long.MAX_VALUE);
+            }
+
+            @Override
+            public void onNext(ByteBuffer item) {
+                digest.update(item.array());
+                items.add(HttpRequest.BodyPublishers.ofByteArray(item.array()));
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                throw new IllegalStateException(throwable);
+            }
+
+            @Override
+            public void onComplete() {
+                payloadHexedSha56 = HEX_FORMATTER.formatHex(digest.digest());
+            }
+        });
+
+        return HttpRequest.BodyPublishers.concat(items.toArray(new BodyPublisher[]{}));
+    }
+
+    /**
+     * This method implements <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html">signs the request</a>
+     * using the AWS v4 algorithm.
+     */
+    private void signRequest() {
+        var isoDateTime = DATE_TIME_FORMATTER.format(date);
+        var isoDate = isoDateTime.substring(0, 8);
+        var unsignedRequest = builder.build();
+
+        var allHeaders = new HashMap<String, String>();
+        unsignedRequest.headers().map().keySet().forEach(k -> unsignedRequest.headers().firstValue(k).ifPresent(value -> allHeaders.put(k, value)));
```

Comment:
Why isn't this `allHeaders = unsignedRequest.headers().map()`?

Reply:
Actually, my implementation is not complete. I'd intended to revisit.   I'm not handing headers with no values or headers with multiple values.  Unfortunately, AFAICS, the AWS signing specification doesn't actually state the rules, but you can see from the test what is required.

https://github.com/aws/aws-sdk-java-v2/blob/26bb6dcf058b08f55665f931d02937238b00e576/core/auth/src/test/java/software/amazon/awssdk/auth/signer/Aws4SignerTest.java#L182

OTOH, for our case case, we don't actually need this support, so we could potentially just fail with a NotImplementedException.  WDYT?

Reply:
If we're not handling it/can't test it then failing with an exception seems like the right thing to do.